src/inspector_profiler.cc

```diff
@@ -0,0 +1,216 @@
+    if (!t->NewInstance(context).ToLocal(&obj)) {
+      return MaybeLocal<Object>();
+    }
+
+    obj->SetAlignedPointerInInternalField(0, nullptr);
+    return scope.Escape(obj);
+  }
+
+  void Start() {
+    SetConnection(object());
+    StartProfiling();
+  }
+
+  void End(Local<Function> callback) {
+    SetMessageCallback(callback);
+    EndProfiling();
+  }
+
+  // Override this to return a JS function that gets called with the message
+  // sent from the session.
+  virtual Local<Function> GetMessageCallback() = 0;
+  virtual void SetMessageCallback(Local<Function> callback) = 0;
+  // Use DispatchMessage() to dispatch necessary inspector messages
+  virtual void StartProfiling() = 0;
+  virtual void EndProfiling() = 0;
+  virtual void SetConnection(Local<Object> connection) = 0;
+
+  friend class V8ProfilerSessionDelegate;
+
+ private:
+  std::unique_ptr<inspector::InspectorSession> session_;
+};
+
+class V8CoverageConnection : public V8ProfilerConnection {
+ public:
+  explicit V8CoverageConnection(Environment* env)
+      : V8ProfilerConnection(env,
+                             CreateConnectionObject(env).ToLocalChecked()) {}
+
+  Local<Function> GetMessageCallback() override {
+    return env()->on_coverage_message_function();
+  }
+
+  void SetMessageCallback(Local<Function> callback) override {
+    return env()->set_on_coverage_message_function(callback);
+  }
+
+  static V8ProfilerConnection* GetConnection(Environment* env) {
+    return Unwrap<V8CoverageConnection>(env->coverage_connection());
+  }
+
+  void SetConnection(Local<Object> obj) override {
+    env()->set_coverage_connection(obj);
+  }
+
+  void StartProfiling() override {
+    Debug(this->env(),
+          DebugCategory::INSPECTOR_PROFILER,
+          "Sending Profiler.startPreciseCoverage\n");
+    Isolate* isolate = this->env()->isolate();
+    Local<String> enable = FIXED_ONE_BYTE_STRING(
+        isolate, "{\"id\": 1, \"method\": \"Profiler.enable\"}");
+    Local<String> start = FIXED_ONE_BYTE_STRING(
+        isolate,
+        "{"
+        "\"id\": 2,"
+        "\"method\": \"Profiler.startPreciseCoverage\","
+        "\"params\": {\"callCount\": true, \"detailed\": true}"
+        "}");
+    DispatchMessage(isolate, enable);
+    DispatchMessage(isolate, start);
+  }
+
+  void EndProfiling() override {
+    Debug(this->env(),
+          DebugCategory::INSPECTOR_PROFILER,
+          "Sending Profiler.takePreciseCoverage\n");
+    Isolate* isolate = this->env()->isolate();
+    Local<String> end =
+        FIXED_ONE_BYTE_STRING(isolate,
+                              "{"
+                              "\"id\": 3,"
+                              "\"method\": \"Profiler.takePreciseCoverage\""
+                              "}");
+    DispatchMessage(isolate, end);
+  }
+
+ private:
+  std::unique_ptr<inspector::InspectorSession> session_;
+};
+
+void StartCoverageCollection(Environment* env) {
+  V8CoverageConnection* connection = new V8CoverageConnection(env);
```

Comment:
auto here, https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-auto.html#new-expressions ?

Reply:
In this case our style guide takes precedence because it's a simple type: https://github.com/nodejs/node/blob/master/CPP_STYLE_GUIDE.md#using-auto

> `Being explicit about types is usually preferred over using auto.`